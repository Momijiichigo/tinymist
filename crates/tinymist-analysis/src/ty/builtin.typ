
#import "typings.typ": *

#let T = var();
#let U = var();

/// Creates a new color from the given RGB components.
///
/// - r (int): The R component of the color.
/// - g (int): The G component of the color.
/// - b (int): The B component of the color.
/// -> color
#let rgb(r: int, g: int, b: int) = func;

/// Creates a new color from the given RGB components.
///
/// - hex (str): A hexadecimal representation of the color.
/// -> color
#let rgb2(hex: str) = func.with(alias: rgb);

/// Maps the elements of the array to a new array using the given function.
///
/// - self (array): The array to map.
/// - f (function): The function to apply to each element.
/// -> array
#let map(self: arr(T), f: (elem: T) => U) = func.with(ret: arr(U));
