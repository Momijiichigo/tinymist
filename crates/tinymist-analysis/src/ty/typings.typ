#let func(..args) = metadata(args);
#let var(..args) = metadata(args);
#let arr(..args) = metadata(args);
