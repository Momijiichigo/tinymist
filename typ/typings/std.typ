
#import "/typ/packages/typings/lib.typ": *

#let U = tv("U");
#let V = tv("V");
#let W = tv("W");

#let dict-type(V: V) = {
  let Self = Self.with(V)
  rec(
    name: "dictionary",
    scope: (
      len: (self: Self) => int,
      at: (self: Self, key: pos(str), default: pos(V)) => V,
      insert: (self: Self, key: pos(str), value: pos(V)) => none,
      remove: (self: Self, key: pos(str), default: pos(V)) => none,
      keys: (self: Self) => arr(str),
      values: (self: Self) => arr(V),
      pairs: (self: Self) => arr(tuple(str, V)),
    ),
  )
};

#let array-type(V: V) = {
  let Arr = Self
  let Self = Self.with(V)
  rec(
    name: "array",
    scope: (
      range: (end: pos(int), start: named(int, 0), step: named(int, 1)) => arr(int),
      len: (self: Self) => int,
      first: (self: Self) => V,
      last: (self: Self) => V,
      at: (self: Self, index: pos(int), default: pos(V)) => V,
      push: (self: Self, value: pos(V)) => none,
      insert: (self: Self, index: pos(int), value: pos(V)) => none,
      remove: (self: Self, index: pos(str), default: pos(V)) => none,
      slice: (self: Self, start: pos(int), end: pos(int), count: pos(int)) => arr(V),
      contains: (self: Self, value: pos(V)) => bool,
      find: (self: Self, searcher: (elem: pos(V)) => bool) => V,
      position: (self: Self, searcher: (elem: pos(V)) => bool) => int,
      filter: (self: Self, searcher: (elem: pos(V)) => bool) => arr(V),
      map: (self: Self, f: pos((elem: pos(V)) => U)) => arr(U),
      enumerate: (self: Self, start: named(int, 0)) => arr(tuple(int, V)),
      zip: (self: Self, others: pos(arr(U)), exact: named(bool, false)) => arr(tuple(V, U)),
      fold: (
        self: Self,
        init: pos(U),
        folder: pos((acc: pos(U), elem: pos(V)) => U),
      ) => U,
      sum: (self: Self, default: pos(V)) => V,
      product: (self: Self, default: pos(V)) => V,
      any: (self: Self, test: (elem: pos(V)) => bool) => bool,
      all: (self: Self, test: (elem: pos(V)) => bool) => bool,
      flatten: (self: Arr.with(arr(V))) => arr(V),
      rev: (self: Self) => arr(V),
      split: (self: Self, at: pos(V)) => arr(arr(V)),
      join: sig(
        (self: Self, last: named(W)) => add(V, W),
        (self: Self, separator: pos(U), last: named(W)) => add(V, add(U, W)),
      ),
      intersperse: (self: Self, separator: pos(U)) => arr(union(V, U)),
      chunks: (self: Self, size: pos(int), exact: named(bool, false)) => arr(arr(V)),
      windows: (self: Self, window-size: pos(int)) => arr(arr(V)),
      sorted: (
        self: Self,
        key: named((elem: pos(V)) => any),
      ) => arr(V),
      dedup: (
        self: Self,
        key: named((elem: pos(V)) => any),
      ) => arr(V),
      to-dict: (
        self: Arr.with(tuple(str, V)),
      ) => dict(str, V),
      reduce: (
        self: Self,
        reducer: pos((acc: pos(V), elem: pos(V)) => V),
      ) => opt(V),
    ),
  )
};


#let assert = union(
  (condition: pos(bool), message: named(str)) => invariant(condition),
  eq: (left: pos(any), right: pos(any), message: named(str)) => invariant(eq(left, right)),
  ne: (left: pos(any), right: pos(any), message: named(str)) => invariant(neq(left, right)),
);
