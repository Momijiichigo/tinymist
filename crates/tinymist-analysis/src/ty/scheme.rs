#![allow(unused)]

use core::fmt;
use std::{path::Path, sync::OnceLock};

use comemo::Track;
use typst::{
    engine::Engine,
    foundations::{func::Repr, Args, Context, Func, Str, Value},
    syntax::Source,
};

use super::{term_value, Ty};
use crate::ty::{BuiltinTy, Interned, TypeInfo};

pub struct TySchemeWorker<'a> {
    scheme: &'a mut TypeInfo,
}

impl TySchemeWorker<'_> {
    fn is_typing_item(&self, v: &Value) -> bool {
        matches!(v, Value::Func(f) if f.span().id().is_some_and(|s| s.vpath().as_rooted_path() == Path::new("/typings.typ")))
    }
}

impl TySchemeWorker<'_> {
    pub fn define(&mut self, k: &str, v: &Value) -> Ty {
        if !self.is_typing_item(v) {
            self.define_value(v)
        } else {
            self.define_typing_value(k, v)
        }
    }

    fn define_typing_value(&mut self, k: &str, v: &Value) -> Ty {
        let closure = match v {
            Value::Func(f) => f,
            _ => return Ty::Any,
        };

        let with = match closure.inner() {
            Repr::With(c) => c,
            _ => return Ty::Any,
        };

        self.define_typing(k, &with.0, with.1.clone())
    }

    fn define_typing(&mut self, k: &str, func: &Func, mut args: Args) -> Ty {
        match func.inner() {
            Repr::Native(..) | Repr::Element(..) | Repr::Plugin(..) => Ty::Any,
            Repr::With(with) => {
                args.items = with.1.items.iter().cloned().chain(args.items).collect();
                self.define_typing(k, &with.0, args)
            }
            Repr::Closure(closure) => self.ty_cons(k, args).unwrap_or(Ty::Any),
        }
    }

    fn ty_cons(&mut self, k: &str, mut args: Args) -> Option<Ty> {
        crate::log_debug_ct!("ty_cons({k}) args: {args:?}");
        let kind = args.named::<Str>("kind").ok()??;
        Some(match kind.as_str() {
            "var" => self.define(k, &args.eat::<Value>().ok()??),
            "arr" => {
                let ty = self.define(k, &args.eat::<Value>().ok()??);
                Ty::Array(Interned::new(ty))
            }
            "tuple" => {
                let values = args.all::<Value>().ok()?;
                let values = values
                    .into_iter()
                    .map(|v| self.define(k, &v))
                    .collect::<Vec<_>>();
                Ty::Tuple(Interned::new(values))
            }
            _ => Ty::Any,
        })
    }

    fn define_value(&self, v: &Value) -> Ty {
        let ty = term_value(v);
        match ty {
            Ty::Builtin(BuiltinTy::TypeType(ty)) => Ty::Builtin(BuiltinTy::Type(ty)),
            ty => ty,
        }
    }
}

#[cfg(test)]
pub mod tests {
    use std::path::Path;

    use super::*;

    use tinymist_world::{args::CompileOnceArgs, ShadowApi};
    use typst::{
        engine::{Route, Sink, Traced},
        foundations::Bytes,
        introspection::Introspector,
        World,
    };

    use crate::{tests::*, ty::TypeInfo};

    #[test]
    fn test_check() {
        snapshot_testing("type_schema", &|mut world, _path| {
            let main_id = world.main();
            world
                .map_shadow_by_id(
                    main_id.join("/typings.typ"),
                    Bytes::from_string(include_str!("typings.typ")),
                )
                .unwrap();
            world
                .map_shadow_by_id(
                    main_id.join("/builtin.typ"),
                    Bytes::from_string(include_str!("builtin.typ")),
                )
                .unwrap();
            let source = world.source(main_id).unwrap();

            let module = typst_shim::eval::eval_compat(&world, &source).unwrap();

            let mut scheme = TypeInfo::default();
            let mut w = TySchemeWorker {
                scheme: &mut scheme,
            };
            for (k, v) in module.scope().iter() {
                let fid = v.span().id().unwrap();
                if fid != source.id() {
                    continue;
                }

                let ty = w.define(k, v.read());
                w.scheme.exports.insert(k.into(), ty);
            }

            let result = format!("{:#?}", TypeCheckSnapshot(&scheme));

            insta::assert_snapshot!(result);
        });
    }
}
