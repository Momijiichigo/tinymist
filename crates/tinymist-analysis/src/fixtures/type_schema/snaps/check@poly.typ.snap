---
source: crates/tinymist-analysis/src/ty/scheme.rs
expression: result
input_file: crates/tinymist-analysis/src/fixtures/type_schema/poly.typ
---
"rgb" = (((Type(float) | Type(ratio)), (Type(float) | Type(ratio)), (Type(float) | Type(ratio))) => Type(color) | (Type(str)) => Type(color) | (Any) => Type(color))
